### PR TITLE
suppress preconditionviolation when worker unreachable

### PR DESCRIPTION
**Pull request by adiprerepa**
related: https://github.com/codalab/codalab-worksheets/issues/1255

read https://github.com/codalab/codalab-worksheets/issues/1255#issuecomment-826405871

### CODE REVIEW

File: codalab/lib/download_manager.py [hunk 1]
```diff
@@ -361,7 +361,8 @@ def _send_netcat_message(self, worker, response_socket_id, uuid, port, message):
     def _get_read_response_stream(self, response_socket_id):
         with closing(self._worker_model.start_listening(response_socket_id)) as sock:
             header_message = self._worker_model.get_json_message(sock, 60)
-            precondition(header_message is not None, 'Unable to reach worker')
+            if header_message is None:
+                logging.info('Unable to reach worker')
```
Comment (percyliang) [inline codalab/lib/download_manager.py:365]: I'm mildly worried that there's an important case that we're suppressing.  Could we have these go to sentry?

Comment (adiprerepa) [reply codalab/lib/download_manager.py:365]: If we want the error to go to sentry, I think we should leave it as is. Raising `PreconditionViolation` will create a Sentry event

**Comment by epicfaace** [comment]
@adiprerepa sorry, I'm not clear why we need to suppress PreconditionViolation and how that follows from your investigation in #1255. Could you explain?

**Comment by adiprerepa** [comment]
afaik, when you kill the worker, the `socket_id` in the worker table becomes invalid, and when the bundle manager tries to connect, the preconditionviolation is thrown. I'm not really sure if this is expected. wdyt @epicfaace ?

**Comment by epicfaace** [comment]
What does "invalid" mean? Maybe this is expected because the socket is no longer open (since the worker has been killed).

**Comment by adiprerepa** [comment]
right, but should we be throwing an exception for that? or just logging

**Comment by adiprerepa** [comment]
> What does "invalid" mean? Maybe this is expected because the socket is no longer open (since the worker has been killed).

'invalid' means the socket no longer exists because the worker is gone

**Comment by epicfaace** [comment]
> right, but should we be throwing an exception for that? or just logging

@adiprerepa I'm fine with whatever you choose, can you just pick a choice and explain your reasoning for it? The key to this issue (and https://github.com/codalab/codalab-worksheets/issues/1255) is just to pick one of these options and have a solid justification for it that's based on the investigation you already did in https://github.com/codalab/codalab-worksheets/issues/1255.

**Comment by adiprerepa** [comment]
> > right, but should we be throwing an exception for that? or just logging
> 
> @adiprerepa I'm fine with whatever you choose, can you just pick a choice and explain your reasoning for it? The key to this issue (and #1255) is just to pick one of these options and have a solid justification for it that's based on the investigation you already did in #1255.

I think since dead workers are a "fact of life", we shouldn't treat this occurrence any differently than we would if a worker were offline and we tried to send it a message (https://github.com/codalab/codalab-worksheets/blob/master/codalab/lib/download_manager.py#L346). We should just log and move on.

**Comment by epicfaace** [comment]
Makes sense, thanks! Also, why did https://github.com/codalab/codalab-worksheets/pull/1627 not fix this? Did that maybe only silence a subset of the preconditionviolation errors?

**Comment by adiprerepa** [comment]
yeah, it didn't suppress everything

**Comment by adiprerepa** [comment]
this should throw an error in the API I think

**Comment by epicfaace** [comment]
Sounds good. What does the error look like on the API end? (Status code, response body?) Just want to make sure it's what we want.

**Comment by adiprerepa** [comment]
reach out to user to see if `PreconditionViolation` is impacting their experience

**Comment by adiprerepa** [comment]
So this isn't good
image
Functionality seems to be affected when a `PreconditionViolation` happens, unless this is a one-off

**Comment by adiprerepa** [comment]
yeah, I can't reach this bundle, so can't reproduce.

**Pull request closed without merging**